From d45122d05c425fb25d38bdb21b10fd2f8c4a412b Mon Sep 17 00:00:00 2001
From: Andrew Jorgensen <ajorgens@amazon.com>
Date: Mon, 9 Sep 2013 09:46:41 -0700
Subject: [PATCH] Make the power-state-change module more forgiving

This module only worked if the delay was explicitly a string in the config,
and the log object wasn't passed in, so it couldn't actually log the exception.

Reviewed-by: Ethan Faust <efaust@amazon.com>
Reviewed-by: Matt Nierzwicki <nierzwic@amazon.com>
---
 cloudinit/config/cc_power_state_change.py | 37 ++++++++++++++++++++++---------
 1 file changed, 27 insertions(+), 10 deletions(-)

diff --git a/cloudinit/config/cc_power_state_change.py b/cloudinit/config/cc_power_state_change.py
index 188047e..c64b8c6 100644
--- a/cloudinit/config/cc_power_state_change.py
+++ b/cloudinit/config/cc_power_state_change.py
@@ -1,8 +1,10 @@
 # vi: ts=4 expandtab
 #
 #    Copyright (C) 2011 Canonical Ltd.
+#    Copyright (C) 2014 Amazon.com, Inc. or its affiliates.
 #
 #    Author: Scott Moser <scott.moser@canonical.com>
+#    Author: Andrew Jorgensen <ajorgens@amazon.com>
 #
 #    This program is free software: you can redistribute it and/or modify
 #    it under the terms of the GNU General Public License version 3, as
@@ -33,12 +35,12 @@ EXIT_FAIL = 254
 def handle(_name, cfg, _cloud, log, _args):
 
     try:
-        (args, timeout) = load_power_state(cfg)
+        (args, timeout) = load_power_state(cfg, log)
         if args is None:
             log.debug("no power_state provided. doing nothing")
             return
-    except Exception as e:
-        log.warn("%s Not performing power state change!" % str(e))
+    except Exception:
+        log.warning("Not performing power state change!", exc_info=True)
         return
 
     mypid = os.getpid()
@@ -56,7 +58,7 @@ def handle(_name, cfg, _cloud, log, _args):
                  [args, devnull_fp])
 
 
-def load_power_state(cfg):
+def load_power_state(cfg, log):
     # returns a tuple of shutdown_command, timeout
     # shutdown_command is None if no config found
     pstate = cfg.get('power_state')
@@ -69,18 +71,33 @@ def load_power_state(cfg):
 
     opt_map = {'halt': '-H', 'poweroff': '-P', 'reboot': '-r'}
 
-    mode = pstate.get("mode")
+    mode = util.get_cfg_option_str(pstate, 'mode')
+    log.debug('mode: %s', mode)
     if mode not in opt_map:
         raise TypeError("power_state[mode] required, must be one of: %s." %
                         ','.join(opt_map.keys()))
 
-    delay = pstate.get("delay", "now")
-    if delay != "now" and not re.match(r"\+[0-9]+", delay):
-        raise TypeError("power_state[delay] must be 'now' or '+m' (minutes).")
+    # Cast to a str so that there's no confusion below.
+    delay = util.get_cfg_option_str(pstate, 'delay', 'now')
+    log.debug('delay: %s', delay)
+    # Valid values for the delay are now or +m, where m is a number of minutes
+    # from now. now is an alias for +0. shutdown also accepts absolute time in
+    # hh:mm, but this doesn't make sense in this context, where it is not known
+    # what time it will be when the module is run.
+    if delay != 'now':
+        if not re.match(r'\+?[0-9]+', delay):
+            raise TypeError(
+                'power_state[delay] must be \'now\' or \'+m\' (minutes).')
+        # the yaml parser consumes the + and makes it a positive integer, so we
+        # try to be more forgiving here.
+        if delay[0] != '+':
+            delay = '+%s' % delay
 
     args = ["shutdown", opt_map[mode], delay]
-    if pstate.get("message"):
-        args.append(pstate.get("message"))
+    message = util.get_cfg_option_str(pstate, 'message')
+    if message:
+        args.append(message)
+    log.debug('message: %s', message)
 
     try:
         timeout = float(pstate.get('timeout', 30.0))
-- 
1.8.3.1

